File: ts/src/MusicbrainzSDK.ts
// Musicbrainz Ts SDK

import { AreaEntity } from './entity/AreaEntity'
import { ArtistEntity } from './entity/ArtistEntity'
import { CollectionEntity } from './entity/CollectionEntity'
import { EventEntity } from './entity/EventEntity'
import { GenreEntity } from './entity/GenreEntity'
import { InstrumentEntity } from './entity/InstrumentEntity'
import { LabelEntity } from './entity/LabelEntity'
import { PlaceEntity } from './entity/PlaceEntity'
import { RatingEntity } from './entity/RatingEntity'
import { RecordingEntity } from './entity/RecordingEntity'
import { RecordingListEntity } from './entity/RecordingListEntity'
import { ReleaseEntity } from './entity/ReleaseEntity'
import { ReleaseGroupEntity } from './entity/ReleaseGroupEntity'
import { ReleaseListEntity } from './entity/ReleaseListEntity'
import { SeriesEntity } from './entity/SeriesEntity'
import { TagEntity } from './entity/TagEntity'
import { UrlEntity } from './entity/UrlEntity'
import { WorkEntity } from './entity/WorkEntity'
import { WorkListEntity } from './entity/WorkListEntity'

export type * from './MusicbrainzTypes'


import { inspect } from 'node:util'

import type { Context, Feature } from './types'

import { config } from './Config'
import { MusicbrainzEntityBase } from './MusicbrainzEntityBase'
import { Utility } from './utility/Utility'


import { BaseFeature } from './feature/base/BaseFeature'


const stdutil = new Utility()


class MusicbrainzSDK {
  _mode: string = 'live'
  _options: any
  _utility = new Utility()
  _features: Feature[]
  _rootctx: Context

  constructor(options?: any) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath
    const items = struct.items

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    items(this._options.feature, (fitem: [string, any]) => {
      const fname = fitem[0]
      const fopts = fitem[1]
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    })

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs?: any) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx: Context = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec: any = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  async direct(fetchargs?: any) {
    const utility = this._utility
    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx: Context = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json: any = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err: any) {
      return { ok: false, err }
    }
  }



  _area?: AreaEntity

  // Idiomatic facade: `client.area.list()` / `client.area.load({ id })`.
  get area(): AreaEntity {
    return (this._area ??= new AreaEntity(this, undefined))
  }

  /** @deprecated Use `client.area` instead. */
  Area(data?: any) {
    const self = this
    return new AreaEntity(self,data)
  }


  _artist?: ArtistEntity

  // Idiomatic facade: `client.artist.list()` / `client.artist.load({ id })`.
  get artist(): ArtistEntity {
    return (this._artist ??= new ArtistEntity(this, undefined))
  }

  /** @deprecated Use `client.artist` instead. */
  Artist(data?: any) {
    const self = this
    return new ArtistEntity(self,data)
  }


  _collection?: CollectionEntity

  // Idiomatic facade: `client.collection.list()` / `client.collection.load({ id })`.
  get collection(): CollectionEntity {
    return (this._collection ??= new CollectionEntity(this, undefined))
  }

  /** @deprecated Use `client.collection` instead. */
  Collection(data?: any) {
    const self = this
    return new CollectionEntity(self,data)
  }


  _event?: EventEntity

  // Idiomatic facade: `client.event.list()` / `client.event.load({ id })`.
  get event(): EventEntity {
    return (this._event ??= new EventEntity(this, undefined))
  }

  /** @deprecated Use `client.event` instead. */
  Event(data?: any) {
    const self = this
    return new EventEntity(self,data)
  }


  _genre?: GenreEntity

  // Idiomatic facade: `client.genre.list()` / `client.genre.load({ id })`.
  get genre(): GenreEntity {
    return (this._genre ??= new GenreEntity(this, undefined))
  }

  /** @deprecated Use `client.genre` instead. */
  Genre(data?: any) {
    const self = this
    return new GenreEntity(self,data)
  }


  _instrument?: InstrumentEntity

  // Idiomatic facade: `client.instrument.list()` / `client.instrument.load({ id })`.
  get instrument(): InstrumentEntity {
    return (this._instrument ??= new InstrumentEntity(this, undefined))
  }

  /** @deprecated Use `client.instrument` instead. */
  Instrument(data?: any) {
    const self = this
    return new InstrumentEntity(self,data)
  }


  _label?: LabelEntity

  // Idiomatic facade: `client.label.list()` / `client.label.load({ id })`.
  get label(): LabelEntity {
    return (this._label ??= new LabelEntity(this, undefined))
  }

  /** @deprecated Use `client.label` instead. */
  Label(data?: any) {
    const self = this
    return new LabelEntity(self,data)
  }


  _place?: PlaceEntity

  // Idiomatic facade: `client.place.list()` / `client.place.load({ id })`.
  get place(): PlaceEntity {
    return (this._place ??= new PlaceEntity(this, undefined))
  }

  /** @deprecated Use `client.place` instead. */
  Place(data?: any) {
    const self = this
    return new PlaceEntity(self,data)
  }


  _rating?: RatingEntity

  // Idiomatic facade: `client.rating.list()` / `client.rating.load({ id })`.
  get rating(): RatingEntity {
    return (this._rating ??= new RatingEntity(this, undefined))
  }

  /** @deprecated Use `client.rating` instead. */
  Rating(data?: any) {
    const self = this
    return new RatingEntity(self,data)
  }


  _recording?: RecordingEntity

  // Idiomatic facade: `client.recording.list()` / `client.recording.load({ id })`.
  get recording(): RecordingEntity {
    return (this._recording ??= new RecordingEntity(this, undefined))
  }

  /** @deprecated Use `client.recording` instead. */
  Recording(data?: any) {
    const self = this
    return new RecordingEntity(self,data)
  }


  _recording_list?: RecordingListEntity

  // Idiomatic facade: `client.recording_list.list()` / `client.recording_list.load({ id })`.
  get recording_list(): RecordingListEntity {
    return (this._recording_list ??= new RecordingListEntity(this, undefined))
  }

  /** @deprecated Use `client.recording_list` instead. */
  RecordingList(data?: any) {
    const self = this
    return new RecordingListEntity(self,data)
  }


  _release?: ReleaseEntity

  // Idiomatic facade: `client.release.list()` / `client.release.load({ id })`.
  get release(): ReleaseEntity {
    return (this._release ??= new ReleaseEntity(this, undefined))
  }

  /** @deprecated Use `client.release` instead. */
  Release(data?: any) {
    const self = this
    return new ReleaseEntity(self,data)
  }


  _release_group?: ReleaseGroupEntity

  // Idiomatic facade: `client.release_group.list()` / `client.release_group.load({ id })`.
  get release_group(): ReleaseGroupEntity {
    return (this._release_group ??= new ReleaseGroupEntity(this, undefined))
  }

  /** @deprecated Use `client.release_group` instead. */
  ReleaseGroup(data?: any) {
    const self = this
    return new ReleaseGroupEntity(self,data)
  }


  _release_list?: ReleaseListEntity

  // Idiomatic facade: `client.release_list.list()` / `client.release_list.load({ id })`.
  get release_list(): ReleaseListEntity {
    return (this._release_list ??= new ReleaseListEntity(this, undefined))
  }

  /** @deprecated Use `client.release_list` instead. */
  ReleaseList(data?: any) {
    const self = this
    return new ReleaseListEntity(self,data)
  }


  _series?: SeriesEntity

  // Idiomatic facade: `client.series.list()` / `client.series.load({ id })`.
  get series(): SeriesEntity {
    return (this._series ??= new SeriesEntity(this, undefined))
  }

  /** @deprecated Use `client.series` instead. */
  Series(data?: any) {
    const self = this
    return new SeriesEntity(self,data)
  }


  _tag?: TagEntity

  // Idiomatic facade: `client.tag.list()` / `client.tag.load({ id })`.
  get tag(): TagEntity {
    return (this._tag ??= new TagEntity(this, undefined))
  }

  /** @deprecated Use `client.tag` instead. */
  Tag(data?: any) {
    const self = this
    return new TagEntity(self,data)
  }


  _url?: UrlEntity

  // Idiomatic facade: `client.url.list()` / `client.url.load({ id })`.
  get url(): UrlEntity {
    return (this._url ??= new UrlEntity(this, undefined))
  }

  /** @deprecated Use `client.url` instead. */
  Url(data?: any) {
    const self = this
    return new UrlEntity(self,data)
  }


  _work?: WorkEntity

  // Idiomatic facade: `client.work.list()` / `client.work.load({ id })`.
  get work(): WorkEntity {
    return (this._work ??= new WorkEntity(this, undefined))
  }

  /** @deprecated Use `client.work` instead. */
  Work(data?: any) {
    const self = this
    return new WorkEntity(self,data)
  }


  _work_list?: WorkListEntity

  // Idiomatic facade: `client.work_list.list()` / `client.work_list.load({ id })`.
  get work_list(): WorkListEntity {
    return (this._work_list ??= new WorkListEntity(this, undefined))
  }

  /** @deprecated Use `client.work_list` instead. */
  WorkList(data?: any) {
    const self = this
    return new WorkListEntity(self,data)
  }




  static test(testoptsarg?: any, sdkoptsarg?: any) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new MusicbrainzSDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts?: any, sdkopts?: any) {
    return MusicbrainzSDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'Musicbrainz' }
  }

  toString() {
    return 'Musicbrainz ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = MusicbrainzSDK


export {
  stdutil,

  BaseFeature,
  MusicbrainzEntityBase,

  MusicbrainzSDK,
  SDK,
}

File: lua/musicbrainz_sdk.lua
-- Musicbrainz SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Typed-model annotations (LuaLS ---@class); empty at runtime.
require("musicbrainz_types")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local MusicbrainzSDK = {}
MusicbrainzSDK.__index = MusicbrainzSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

MusicbrainzSDK._make_feature = _make_feature


function MusicbrainzSDK.new(options)
  local self = setmetatable({}, MusicbrainzSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config_shared")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features in the resolved order (make_options puts an explicit list
  -- order first, else defaults to test-first). Ordering matters: the `test`
  -- feature installs the base mock transport and the transport features
  -- (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
  -- must be added before them to sit at the base of the chain.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local featureorder = vs.getpath(self.options, "__derived__.featureorder")
    if type(featureorder) == "table" then
      for _, fname in ipairs(featureorder) do
        local fopts = helpers.to_map(feature_opts[fname])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

    -- feature: test


  return self
end


function MusicbrainzSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function MusicbrainzSDK:get_utility()
  return Utility.copy(self._utility)
end


function MusicbrainzSDK:get_root_ctx()
  return self._rootctx
end


function MusicbrainzSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


-- Raw endpoint access is operator-controllable, like every entity op.
-- Blocking it means denying BOTH the 'direct' and 'graphql' tokens, since
-- either one reaches the same endpoint.
function MusicbrainzSDK:direct(fetchargs)
  if not self:_op_allowed("direct") then
    return self:_op_denied("direct"), nil
  end

  return self:_raw_request(fetchargs)
end


-- Is this raw-access op permitted by the SDK's allow.op option?
function MusicbrainzSDK:_op_allowed(op)
  local allow = vs.getpath(self.options, "allow.op")
  return type(allow) == "string" and allow:find(op, 1, true) ~= nil
end


function MusicbrainzSDK:_op_denied(op)
  local allow = vs.getpath(self.options, "allow.op")
  if type(allow) ~= "string" then allow = "" end
  return {
    ok = false,
    err = "MusicbrainzSDK: " .. op .. ": operation not allowed by" ..
      " SDK option allow.op value: \"" .. allow .. "\"",
  }
end


-- Ungated request path shared by direct and graphql, each of which checks its
-- own allow.op token first. Private, rather than a flag on fetchargs: a
-- caller-supplied marker would let anyone opt straight back out of the gate
-- by passing it.
function MusicbrainzSDK:_raw_request(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end


-- Raw GraphQL access: the pressure valve that makes the generated surface's
-- deliberate omissions (per-call selection sets, typed filter builders,
-- batching, subscriptions) livable — the whole schema stays reachable.
--
-- Thin wrapper over the same prepare/fetch path direct uses, with the one
-- thing raw direct cannot do for GraphQL: a GraphQL failure rides HTTP 200 as
-- a top-level `errors` array, so status alone would report a failed query as
-- ok.
--
-- NOTE: like direct, this bypasses the feature pipeline — no retry, ratelimit
-- or paging features apply.
function MusicbrainzSDK:graphql(query, variables, ctrl)
  if not self:_op_allowed("graphql") then
    return self:_op_denied("graphql"), nil
  end

  local res, err = self:_raw_request({
    method = "POST",
    headers = { ["content-type"] = "application/json" },
    body = {
      query = query,
      variables = type(variables) == "table" and variables or {},
    },
    ctrl = type(ctrl) == "table" and ctrl or {},
  })

  if err ~= nil or type(res) ~= "table" then
    return res, err
  end

  -- Errors are read BEFORE any status check: a GraphQL parse or validation
  -- failure comes back as HTTP 400 carrying the standard { errors = {...} }
  -- body, and the raw path represents a non-2xx as ok=false with no err — so
  -- returning early on status would discard the server's own diagnostics,
  -- which are the only useful part of that response.
  local errors = vs.getpath(res, "data.errors")

  if type(errors) == "table" and 0 < #errors then
    local msg = vs.getprop(errors[1], "message")
    if type(msg) ~= "string" or msg == "" then
      msg = "graphql error"
    end
    res.ok = false
    res.err = "MusicbrainzSDK: graphql: " .. msg
    res.graphql = errors
  end

  return res, nil
end



-- Idiomatic facade: client:Area():list() / client:Area():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Area(data)
  local EntityMod = require("entity.area_entity")
  if data == nil then
    if self._area == nil then
      self._area = EntityMod.new(self, nil)
    end
    return self._area
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Artist():list() / client:Artist():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Artist(data)
  local EntityMod = require("entity.artist_entity")
  if data == nil then
    if self._artist == nil then
      self._artist = EntityMod.new(self, nil)
    end
    return self._artist
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Collection():list() / client:Collection():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Collection(data)
  local EntityMod = require("entity.collection_entity")
  if data == nil then
    if self._collection == nil then
      self._collection = EntityMod.new(self, nil)
    end
    return self._collection
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Event():list() / client:Event():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Event(data)
  local EntityMod = require("entity.event_entity")
  if data == nil then
    if self._event == nil then
      self._event = EntityMod.new(self, nil)
    end
    return self._event
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Genre():list() / client:Genre():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Genre(data)
  local EntityMod = require("entity.genre_entity")
  if data == nil then
    if self._genre == nil then
      self._genre = EntityMod.new(self, nil)
    end
    return self._genre
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Instrument():list() / client:Instrument():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Instrument(data)
  local EntityMod = require("entity.instrument_entity")
  if data == nil then
    if self._instrument == nil then
      self._instrument = EntityMod.new(self, nil)
    end
    return self._instrument
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Label():list() / client:Label():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Label(data)
  local EntityMod = require("entity.label_entity")
  if data == nil then
    if self._label == nil then
      self._label = EntityMod.new(self, nil)
    end
    return self._label
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Place():list() / client:Place():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Place(data)
  local EntityMod = require("entity.place_entity")
  if data == nil then
    if self._place == nil then
      self._place = EntityMod.new(self, nil)
    end
    return self._place
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Rating():list() / client:Rating():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Rating(data)
  local EntityMod = require("entity.rating_entity")
  if data == nil then
    if self._rating == nil then
      self._rating = EntityMod.new(self, nil)
    end
    return self._rating
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Recording():list() / client:Recording():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Recording(data)
  local EntityMod = require("entity.recording_entity")
  if data == nil then
    if self._recording == nil then
      self._recording = EntityMod.new(self, nil)
    end
    return self._recording
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:RecordingList():list() / client:RecordingList():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:RecordingList(data)
  local EntityMod = require("entity.recording_list_entity")
  if data == nil then
    if self._recording_list == nil then
      self._recording_list = EntityMod.new(self, nil)
    end
    return self._recording_list
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Release():list() / client:Release():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Release(data)
  local EntityMod = require("entity.release_entity")
  if data == nil then
    if self._release == nil then
      self._release = EntityMod.new(self, nil)
    end
    return self._release
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ReleaseGroup():list() / client:ReleaseGroup():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:ReleaseGroup(data)
  local EntityMod = require("entity.release_group_entity")
  if data == nil then
    if self._release_group == nil then
      self._release_group = EntityMod.new(self, nil)
    end
    return self._release_group
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ReleaseList():list() / client:ReleaseList():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:ReleaseList(data)
  local EntityMod = require("entity.release_list_entity")
  if data == nil then
    if self._release_list == nil then
      self._release_list = EntityMod.new(self, nil)
    end
    return self._release_list
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Series():list() / client:Series():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Series(data)
  local EntityMod = require("entity.series_entity")
  if data == nil then
    if self._series == nil then
      self._series = EntityMod.new(self, nil)
    end
    return self._series
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Tag():list() / client:Tag():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Tag(data)
  local EntityMod = require("entity.tag_entity")
  if data == nil then
    if self._tag == nil then
      self._tag = EntityMod.new(self, nil)
    end
    return self._tag
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Url():list() / client:Url():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Url(data)
  local EntityMod = require("entity.url_entity")
  if data == nil then
    if self._url == nil then
      self._url = EntityMod.new(self, nil)
    end
    return self._url
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Work():list() / client:Work():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:Work(data)
  local EntityMod = require("entity.work_entity")
  if data == nil then
    if self._work == nil then
      self._work = EntityMod.new(self, nil)
    end
    return self._work
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:WorkList():list() / client:WorkList():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MusicbrainzSDK:WorkList(data)
  local EntityMod = require("entity.work_list_entity")
  if data == nil then
    if self._work_list == nil then
      self._work_list = EntityMod.new(self, nil)
    end
    return self._work_list
  end
  return EntityMod.new(self, data)
end




function MusicbrainzSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = MusicbrainzSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return MusicbrainzSDK
